mayastor/src/bdev/nexus/nexus_child_error_store.rs

```diff
@@ -105,61 +113,138 @@ impl NexusErrStore {
+        let found_op = match record.io_op {
+            io_type::READ => (io_op_flags & NexusErrStore::READ_FLAG) != 0,
+            io_type::WRITE => (io_op_flags & NexusErrStore::WRITE_FLAG) != 0,
+            io_type::UNMAP => (io_op_flags & NexusErrStore::UNMAP_FLAG) != 0,
+            io_type::FLUSH => (io_op_flags & NexusErrStore::FLUSH_FLAG) != 0,
+            io_type::RESET => (io_op_flags & NexusErrStore::RESET_FLAG) != 0,
+            _ => false,
+        };
+
+        if found_op {
+            let found_err = match record.io_error {
+                io_status::FAILED => {
+                    (io_error_flags & NexusErrStore::IO_FAILED_FLAG) != 0
+                }
+                _ => false,
+            };
+            if found_err {
+                return match target_timestamp {
+                    Some(ts) => {
+                        record.timestamp.checked_duration_since(ts).is_some()
+                    }
+                    None => true,
+                };
+            }
+        }
+        false
+    }
+
+    #[inline]
+    fn count_errors(
         &self,
         io_op_flags: u32,
         io_error_flags: u32,
         target_timestamp: Option<Instant>,
     ) -> u32 {
         let mut idx = self.next_record_index;
-        let mut error_count: u32 = 0;
+        let mut err_count: u32 = 0;
 
         for _ in 0 .. self.no_of_records {
             if idx > 0 {
                 idx -= 1;
             } else {
                 idx = self.records.len() - 1;
             }
-            if target_timestamp.is_some()
-                && self.records[idx]
-                    .timestamp
-                    .checked_duration_since(target_timestamp.unwrap())
-                    .is_none()
-            {
-                break; // reached a record older than the wanted timespan
+            if Self::is_wanted_record(
+                &self.records[idx],
+                io_op_flags,
+                io_error_flags,
+                target_timestamp,
+            ) {
+                err_count += 1;
             }
-            let found_op = match self.records[idx].io_op {
-                io_type::READ => (io_op_flags & NexusErrStore::READ_FLAG) != 0,
-                io_type::WRITE => {
-                    (io_op_flags & NexusErrStore::WRITE_FLAG) != 0
-                }
-                io_type::UNMAP => {
-                    (io_op_flags & NexusErrStore::UNMAP_FLAG) != 0
-                }
-                io_type::FLUSH => {
-                    (io_op_flags & NexusErrStore::FLUSH_FLAG) != 0
-                }
-                io_type::RESET => {
-                    (io_op_flags & NexusErrStore::RESET_FLAG) != 0
-                }
-                _ => false,
-            };
+        }
+        err_count
+    }
 
-            let found_err = match self.records[idx].io_error {
-                io_status::FAILED => {
-                    (io_error_flags & NexusErrStore::IO_FAILED_FLAG) != 0
+    #[inline]
```

Comment:
I'm not quite sure what we are trying to do here...  If we want to count the errors for a particular type, this seems rather expensive.

Reply:
count_repeats() now taken out of this PR as a result of taking out the repeated IO error handling, as suggested in this review.